From 23e88e260d2d90e88ee03fd6e9d8e42265dcba8d Mon Sep 17 00:00:00 2001
From: Aaron Bassett <abassett@qnx.com>
Date: Wed, 4 Mar 2026 17:00:40 +0000
Subject: [PATCH] qnx support; add x86_64 support for qnx

Imported from qnx-ports/nspr (396c53cad4db "release tarball for 4.38.2"
...23e88e260d2d, branch qnx-4.38.x); original authorship kept. Applies to
4.39 unchanged (offsets only).

NSPR's NTO (QNX) machine layer dates from QNX 6 on x86: its _nto.cfg knows
only i386 and _nto.h reports architecture "x86", disables poll() and large
files. QNX 8.0 has poll(), 64-bit off_t and the 5-argument getprotobyname_r /
int-returning gethostbyname_r; this adds the aarch64 and x86_64 type layouts,
enables poll and large files, includes <spawn.h> for posix_spawn, and selects
the reentrant netdb variants QNX provides.

Upstream-Status: QNX-specific (carried in the qnx-ports fork)
---
diff --git a/nspr/pr/include/md/_nto.cfg b/nspr/pr/include/md/_nto.cfg
index b76a560..cbf4f00 100644
--- a/nspr/pr/include/md/_nto.cfg
+++ b/nspr/pr/include/md/_nto.cfg
@@ -62,6 +62,154 @@
 #define PR_BYTES_PER_DWORD_LOG2  3L
 #define PR_WORDS_PER_DWORD_LOG2  1L
 
+#elif defined(__aarch64__)
+
+#ifdef __AARCH64EB__
+#undef  IS_LITTLE_ENDIAN
+#define IS_BIG_ENDIAN 1
+#elif defined(__AARCH64EL__)
+#define IS_LITTLE_ENDIAN 1
+#undef  IS_BIG_ENDIAN
+#else
+#error "Unknown Aarch64 endianness."
+#endif
+#define IS_64
+
+#define PR_BYTES_PER_BYTE   1
+#define PR_BYTES_PER_SHORT  2
+#define PR_BYTES_PER_INT    4
+#define PR_BYTES_PER_INT64  8
+#define PR_BYTES_PER_LONG   8
+#define PR_BYTES_PER_FLOAT  4
+#define PR_BYTES_PER_DOUBLE 8
+#define PR_BYTES_PER_WORD   8
+#define PR_BYTES_PER_DWORD  8
+
+#define PR_BITS_PER_BYTE    8
+#define PR_BITS_PER_SHORT   16
+#define PR_BITS_PER_INT     32
+#define PR_BITS_PER_INT64   64
+#define PR_BITS_PER_LONG    64
+#define PR_BITS_PER_FLOAT   32
+#define PR_BITS_PER_DOUBLE  64
+#define PR_BITS_PER_WORD    64
+
+#define PR_BITS_PER_BYTE_LOG2   3
+#define PR_BITS_PER_SHORT_LOG2  4
+#define PR_BITS_PER_INT_LOG2    5
+#define PR_BITS_PER_INT64_LOG2  6
+#define PR_BITS_PER_LONG_LOG2   6
+#define PR_BITS_PER_FLOAT_LOG2  5
+#define PR_BITS_PER_DOUBLE_LOG2 6
+#define PR_BITS_PER_WORD_LOG2   6
+
+#define PR_ALIGN_OF_SHORT   2
+#define PR_ALIGN_OF_INT     4
+#define PR_ALIGN_OF_LONG    8
+#define PR_ALIGN_OF_INT64   8
+#define PR_ALIGN_OF_FLOAT   4
+#define PR_ALIGN_OF_DOUBLE  8
+#define PR_ALIGN_OF_POINTER 8
+#define PR_ALIGN_OF_WORD    8
+
+#define PR_BYTES_PER_WORD_LOG2  3
+#define PR_BYTES_PER_DWORD_LOG2 3
+
+#elif defined(__x86_64__)
+
+#ifdef __ILP32__
+
+#define IS_LITTLE_ENDIAN 1
+#undef  IS_BIG_ENDIAN
+
+#define PR_BYTES_PER_BYTE   1
+#define PR_BYTES_PER_SHORT  2
+#define PR_BYTES_PER_INT    4
+#define PR_BYTES_PER_INT64  8
+#define PR_BYTES_PER_LONG   4
+#define PR_BYTES_PER_FLOAT  4
+#define PR_BYTES_PER_DOUBLE 8
+#define PR_BYTES_PER_WORD   4
+#define PR_BYTES_PER_DWORD  8
+
+#define PR_BITS_PER_BYTE    8
+#define PR_BITS_PER_SHORT   16
+#define PR_BITS_PER_INT     32
+#define PR_BITS_PER_INT64   64
+#define PR_BITS_PER_LONG    32
+#define PR_BITS_PER_FLOAT   32
+#define PR_BITS_PER_DOUBLE  64
+#define PR_BITS_PER_WORD    32
+
+#define PR_BITS_PER_BYTE_LOG2   3
+#define PR_BITS_PER_SHORT_LOG2  4
+#define PR_BITS_PER_INT_LOG2    5
+#define PR_BITS_PER_INT64_LOG2  6
+#define PR_BITS_PER_LONG_LOG2   5
+#define PR_BITS_PER_FLOAT_LOG2  5
+#define PR_BITS_PER_DOUBLE_LOG2 6
+#define PR_BITS_PER_WORD_LOG2   5
+
+#define PR_ALIGN_OF_SHORT   2
+#define PR_ALIGN_OF_INT     4
+#define PR_ALIGN_OF_LONG    4
+#define PR_ALIGN_OF_INT64   8
+#define PR_ALIGN_OF_FLOAT   4
+#define PR_ALIGN_OF_DOUBLE  8
+#define PR_ALIGN_OF_POINTER 4
+#define PR_ALIGN_OF_WORD    4
+
+#define PR_BYTES_PER_WORD_LOG2   2
+#define PR_BYTES_PER_DWORD_LOG2  3
+
+#else
+
+#define IS_LITTLE_ENDIAN 1
+#undef  IS_BIG_ENDIAN
+#define IS_64
+
+#define PR_BYTES_PER_BYTE   1
+#define PR_BYTES_PER_SHORT  2
+#define PR_BYTES_PER_INT    4
+#define PR_BYTES_PER_INT64  8
+#define PR_BYTES_PER_LONG   8
+#define PR_BYTES_PER_FLOAT  4
+#define PR_BYTES_PER_DOUBLE 8
+#define PR_BYTES_PER_WORD   8
+#define PR_BYTES_PER_DWORD  8
+
+#define PR_BITS_PER_BYTE    8
+#define PR_BITS_PER_SHORT   16
+#define PR_BITS_PER_INT     32
+#define PR_BITS_PER_INT64   64
+#define PR_BITS_PER_LONG    64
+#define PR_BITS_PER_FLOAT   32
+#define PR_BITS_PER_DOUBLE  64
+#define PR_BITS_PER_WORD    64
+
+#define PR_BITS_PER_BYTE_LOG2   3
+#define PR_BITS_PER_SHORT_LOG2  4
+#define PR_BITS_PER_INT_LOG2    5
+#define PR_BITS_PER_INT64_LOG2  6
+#define PR_BITS_PER_LONG_LOG2   6
+#define PR_BITS_PER_FLOAT_LOG2  5
+#define PR_BITS_PER_DOUBLE_LOG2 6
+#define PR_BITS_PER_WORD_LOG2   6
+
+#define PR_ALIGN_OF_SHORT   2
+#define PR_ALIGN_OF_INT     4
+#define PR_ALIGN_OF_LONG    8
+#define PR_ALIGN_OF_INT64   8
+#define PR_ALIGN_OF_FLOAT   4
+#define PR_ALIGN_OF_DOUBLE  8
+#define PR_ALIGN_OF_POINTER 8
+#define PR_ALIGN_OF_WORD    8
+
+#define PR_BYTES_PER_WORD_LOG2  3
+#define PR_BYTES_PER_DWORD_LOG2 3
+
+#endif
+
 #else
 
 #error Undefined CPU Architecture
diff --git a/nspr/pr/include/md/_nto.h b/nspr/pr/include/md/_nto.h
index 1094dd1..73a43d4 100644
--- a/nspr/pr/include/md/_nto.h
+++ b/nspr/pr/include/md/_nto.h
@@ -11,7 +11,15 @@
 */
 #define PR_LINKER_ARCH      "nto"
 #define _PR_SI_SYSNAME      "NTO"
-#define _PR_SI_ARCHITECTURE "x86"
+#if defined(__x86_64__)
+#define _PR_SI_ARCHITECTURE "x86-64"
+#elif defined(__aarch64__)
+#define _PR_SI_ARCHITECTURE "aarch64"
+#else
+#error "Unknown CPU architecture"
+#endif
+
+
 #define PR_DLL_SUFFIX       ".so"
 
 #define _PR_VMBASE      0x30000000
@@ -24,12 +32,12 @@
 #define HAVE_WEAK_IO_SYMBOLS
 #endif
 
-#undef  _PR_POLL_AVAILABLE
-#undef  _PR_USE_POLL
+#define  _PR_POLL_AVAILABLE
+#define  _PR_USE_POLL
 #define _PR_HAVE_SOCKADDR_LEN
 #undef  HAVE_BSD_FLOCK
 #define HAVE_FCNTL_FILE_LOCKING
-#define _PR_NO_LARGE_FILES
+#define _PR_HAVE_LARGE_OFF_T
 #define _PR_STAT_HAS_ONLY_ST_ATIME
 #define PR_HAVE_POSIX_NAMED_SHARED_MEMORY
 #define _PR_HAVE_POSIX_SEMAPHORES
@@ -167,6 +175,8 @@ struct _MDCPU {
 #define _MD_IOQ_LOCK()
 #define _MD_IOQ_UNLOCK()
 
+extern void _MD_EarlyInit(void);
+
 #define _MD_INTERVAL_USE_GTOD
 #define _MD_EARLY_INIT          _MD_EarlyInit
 #define _MD_FINAL_INIT          _PR_UnixInit
diff --git a/nspr/pr/src/md/unix/uxproces.c b/nspr/pr/src/md/unix/uxproces.c
index 1ea075f..03bc90d 100644
--- a/nspr/pr/src/md/unix/uxproces.c
+++ b/nspr/pr/src/md/unix/uxproces.c
@@ -23,6 +23,10 @@
 PR_IMPORT_DATA(char**) environ;
 #endif
 
+#if defined(NTO)
+#  include <spawn.h>
+#endif
+
 /*
  * HP-UX 9 doesn't have the SA_RESTART flag.
  */
diff --git a/nspr/pr/src/misc/prnetdb.c b/nspr/pr/src/misc/prnetdb.c
index 3438e30..cc65cf3 100644
--- a/nspr/pr/src/misc/prnetdb.c
+++ b/nspr/pr/src/misc/prnetdb.c
@@ -83,6 +83,13 @@ PRLock* _pr_dnsLock = NULL;
 #  define _PR_HAVE_5_ARG_GETPROTO_R
 #endif
 
+#if defined(NTO)
+#  define _PR_HAVE_GETPROTO_R
+#  define _PR_HAVE_5_ARG_GETPROTO_R
+#  define _PR_HAVE_GETPROTO_R_INT
+#  define _PR_HAVE_GETHOST_R_INT
+#endif
+
 #if !defined(_PR_HAVE_GETPROTO_R)
         PRLock* _getproto_lock = NULL;
 #endif
@@ -596,7 +603,6 @@ static PRStatus CopyProtoent(struct protoent* from, char* buf, PRIntn bufsize,
  * #################################################################
  */
 #if defined(_PR_HAVE_GETHOST_R_INT)
-
 #  define GETHOSTBYNAME(name) \
     (gethostbyname_r(name, &tmphe, tmpbuf, bufsize, &h, &h_err), h)
 #  define GETHOSTBYNAME2(name, af) \
@@ -605,7 +611,6 @@ static PRStatus CopyProtoent(struct protoent* from, char* buf, PRIntn bufsize,
     (gethostbyaddr_r(addr, addrlen, af, &tmphe, tmpbuf, bufsize, &h, &h_err), h)
 
 #elif defined(_PR_HAVE_GETHOST_R_POINTER)
-
 #  define GETHOSTBYNAME(name) \
     gethostbyname_r(name, &tmphe, tmpbuf, bufsize, &h_err)
 #  define GETHOSTBYNAME2(name, af) \
@@ -1169,7 +1174,7 @@ PR_GetProtoByName(const char* name, char* buffer, PRInt32 buflen,
     PR_SetError(PR_DIRECTORY_LOOKUP_ERROR, _MD_ERRNO());
     return PR_FAILURE;
   }
-#elif defined(_PR_HAVE_GETPROTO_R_INT)
+#elif defined(_PR_HAVE_GETPROTO_R_INT) && !defined(_PR_HAVE_5_ARG_GETPROTO_R)
   /*
   ** The buffer needs to be zero'd, and it should be
   ** at least the size of a struct protoent_data.
@@ -1245,7 +1250,7 @@ PR_GetProtoByNumber(PRInt32 number, char* buffer, PRInt32 buflen,
     return PR_FAILURE;
   }
 
-#elif defined(_PR_HAVE_GETPROTO_R_INT)
+#elif defined(_PR_HAVE_GETPROTO_R_INT) && !defined(_PR_HAVE_5_ARG_GETPROTO_R)
   /*
   ** The buffer needs to be zero'd for these OS's.
   */
